common-ui-compose/src/main/java/app/tivi/common/compose/insets.kt

```diff
@@ -14,64 +14,410 @@
  */
 
+@file:Suppress("NOTHING_TO_INLINE")
+
 package app.tivi.common.compose
 
 import android.view.View
+import androidx.compose.foundation.layout.height
+import androidx.compose.foundation.layout.width
 import androidx.compose.runtime.Composable
 import androidx.compose.runtime.Providers
-import androidx.compose.runtime.ambientOf
-import androidx.compose.runtime.livedata.observeAsState
+import androidx.compose.runtime.Stable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.onCommit
 import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.runtime.staticAmbientOf
+import androidx.compose.ui.LayoutModifier
+import androidx.compose.ui.Measurable
+import androidx.compose.ui.MeasureScope
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.layout.IntrinsicMeasurable
+import androidx.compose.ui.layout.IntrinsicMeasureScope
+import androidx.compose.ui.platform.ViewAmbient
+import androidx.compose.ui.unit.Constraints
+import androidx.compose.ui.unit.offset
 import androidx.core.view.ViewCompat
-import androidx.core.view.WindowInsetsCompat
-import androidx.core.view.doOnAttach
-import androidx.core.view.doOnDetach
-import androidx.lifecycle.LiveData
-import androidx.lifecycle.MutableLiveData
-import androidx.lifecycle.distinctUntilChanged
-import androidx.lifecycle.map
-
-data class InsetsHolder(
-    val left: Int = 0,
-    val top: Int = 0,
-    val right: Int = 0,
-    val bottom: Int = 0
-) {
-    constructor(insets: WindowInsetsCompat) : this(
-        insets.systemWindowInsetLeft,
-        insets.systemWindowInsetTop,
-        insets.systemWindowInsetRight,
-        insets.systemWindowInsetBottom
-    )
+import kotlin.math.min
 
-    val horizontal get() = right + left
-    val vertical get() = top + bottom
+/**
+ * Main holder of our inset values.
+ */
+@Stable
+class DisplayInsets {
+    val systemBars = Insets()
+    val systemGestures = Insets()
 }
 
-val InsetsAmbient = ambientOf { InsetsHolder() }
+@Stable
+class Insets {
+    var left by mutableStateOf(0)
+        internal set
+    var top by mutableStateOf(0)
+        internal set
+    var right by mutableStateOf(0)
+        internal set
+    var bottom by mutableStateOf(0)
+        internal set
 
+    /**
+     * TODO: doesn't currently work
+     */
+    var visible by mutableStateOf(true)
+        internal set
+}
+
+val DisplayInsetsAmbient = staticAmbientOf<DisplayInsets>()
+
+/**
+ * Setup and provide [DisplayInsets] to [DisplayInsetsAmbient].
+ *
+ * This will automatically listen for [WindowInsetsCompat] on the host view.
+ *
+ * @param consumeInsets Whether to consume the insets which are dispatched to the host view.
+ */
 @Composable
-fun ProvideInsets(
-    liveData: LiveData<WindowInsetsCompat?>,
-    children: @Composable () -> Unit
+fun ProvideDisplayInsets(
+    consumeInsets: Boolean = true,
```

Comment:
Would you ever want to consume one type of insets but not another?

Reply:
Not from API 30 (and soon androidx.core). The `consumeXXXX()` functions have been deprecated, and apps should now return `WindowInsets.CONSUMED` if they want to mark the insets as consumed.

I'll update this impl once the core update has gone out, which should be 12th August.